backupccl: add some span recordings to the backup resumer

adityamaru opened the pull request:
This change adds some initial span recordings to the backup
resumer that might be interesting when inspecting an executing
backup. The processor and KV level requests will be dealt
with in subsequent commits.

Informs: #64992

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/65725)
<!-- Reviewable:end -->

adityamaru commented:
Just wanted to get some eyes on this general approach before I went any further. Primarily, if there are any better ideas of how we would like to organize trace recording protobufs. 
I went with trying to split backup specific, and common protos that can be used by restore and import as well.

Also welcoming any high-level pointers on if we want to be more structured about where we want to add more recording. I've just used my own discretion at the moment to identify interesting phases in the resumer.

miretskiy commented:
Are we trying to be too structured?  For example:
```
resumerSpan.RecordStructured(&jobspb.BulkJobTraceEvent{Message: "writing backup manifest"})
```

We already have a way of tracing string values, don't we?  Just thinking out loud: what are we getting for having
these strongly typed structured traces? Are we planning on doing some sort of analysis at some point (like how many traces have x.y > z? Or is more of a log message type (just strings, possibly formatted)?

adityamaru commented:
> We already have a way of tracing string values, don't we? Just thinking out loud: what are we getting for having
> these strongly typed structured traces? 

We do, but if we only rely on structured recordings then we don't need the span to be verbose. Only verbose spans accumulate free form logs, and knz had recommended sticking to lightweight structured protos to avoid perf impacts.

re: whether we should have these kinds of recordings at all, I believe they would serve as visual cues when pulling a jobs trace to help narrow down the "phase" in which a job is stuck?

miretskiy commented:
I meant types.StringValue -- it's a proto, and it has a "Value" in it.  What do we get when using bulk specific proto messages?

adityamaru commented:
> I meant types.StringValue -- it's a proto, and it has a "Value" in it. What do we get when using bulk specific proto messages?

oh yes! That I'm 100% onboard with. I just made it its own proto for the time being incase I figured we needed more bulk-specific fields in it. Will change it to types.StringValue if that is not the case.

adityamaru commented on pkg/ccl/backupccl/backup.proto:
```diff
@@ -166,3 +166,18 @@ message RestoreProgress {
+  message Spans {
+    repeated roachpb.Span spans = 1 [(gogoproto.nullable) = false];
+  }
+  map<int32, Spans> node_to_spans = 1 [(gogoproto.nullable) = false];
```
Yeah crossed my mind while I was adding this proto. I don't have a very good intuition about what serializing/deserializing an O(ranges) proto costs us. I believe we will need to serialize/deserialize if we pull the trace from a node other than the coordinator node, as well as when we use the existing `payloads_for_span` generator - https://github.com/cockroachdb/cockroach/blob/723ded97a32a7771ddc5b34fa858b3f60374389a/pkg/sql/sem/builtins/generator_builtins.go#L1552.

I thought it would be valuable to know which span is being executed on where, but maybe we can get away with adding the span to the trace at the ExportRequest level, when the node actually serves the export request?

adityamaru committed:
backupccl: add trace recordings to the backup execution

This change adds trace recordings to noteworthy places during
backup exection.

Release note: None

adityamaru commented:
The first 4 commits are from the refactor in #65576, so the last two commits are where the changes have been made.

adityamaru commented on pkg/ccl/backupccl/backup.proto:
```diff
@@ -182,3 +182,21 @@ message BackupProgressTraceEvent {
+    string retryable_error = 5;
+  }
+  message Response {
+    util.hlc.Timestamp resp_received_time = 1 [(gogoproto.nullable) = false];
```
maybe duration would be more useful here?

adityamaru replied:
changed to duration since that seems more useful than a timestamp.

pbardea reviewed (commented):
The events LGTM; same comments from the other tracing PR also apply to this (re: redaction and possible truncation of the error messages)

adityamaru commented on pkg/ccl/backupccl/backup.proto:
```diff
@@ -170,3 +170,15 @@ message RestoreProgress {
   roachpb.Span dataSpan = 3 [(gogoproto.nullable) = false];
 }
+
+message BackupPlanningTraceEvent {
```
done.

adityamaru commented:
> The events LGTM; same comments from the other tracing PR also apply to this (re: redaction and possible truncation of the error messages)

Yup, moved the helper to a utils file in tracing, and passed errors through that.

adityamaru commented:
(not sure if you wanted to have another look at this)

pbardea commented on pkg/util/tracing/utils.go:
```diff
@@ -65,3 +66,14 @@ func MessageToJSONString(msg protoutil.Message, emitDefaults bool) (string, erro
+
+// RedactAndTruncateErrorForTracing redacts the error and truncates the string
+// representation of the error to a fixed length.
+func RedactAndTruncateErrorForTracing(err error) string {
```
nit: Could we just call this `RedactAndTruncateError` since it's already in the `tracing` package (so callers will already see `tracing.RedactAndTruncateError`?

adityamaru replied:
done!

pbardea reviewed (approved):
LGTM with naming nit

adityamaru commented:
Bazel CI is red because of a macOS cross-build failure. When I trigger a run manually it passes but doesn't seem to reflect here. Going to merge since I don't believe it is related to this PR,  will bring it up with dev inf.

bors r=pbardea
